cores/esp32/esp32-hal-uart.c

```diff
@@ -277,26 +278,24 @@ uint32_t uartAvailableForWrite(uart_t* uart)
     return 0x7f - uart->dev->status.txfifo_cnt;
 }
 
-uint8_t uartRead(uart_t* uart)
+uint8_t uartRead(uart_t* uart, uint8_t *data, size_t to)
 {
     if(uart == NULL || uart->queue == NULL) {
         return 0;
     }
-    uint8_t c;
-    if(xQueueReceive(uart->queue, &c, 0)) {
-        return c;
+    if(xQueueReceive(uart->queue, data, to)) {
+        return 1;
```

Comment:
this alters the expected behavior of uartRead() which returns the character that was read from the uart device.

Reply:
@atanisoft mention 0 timeout in original code. It will return 0 even if there's nothing to read in uart. That's weird behavior and breaks [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) per se and should be fixed.

Reply:
while it may be weird and break POLA, it is likely going to break a number of existing libraries which rely on the current method signatures and behaviors of the methods. I know I have a library which uses the UART HAL (I'll be moving it over to IDF API instead soon).

Reply:
how do you determine case if UART empty or not in your library?

Moving to idf API was the first solution we've considered. But it's not perfect too. If you carefully  look into uart api implementation you'll notice that it's impossible to set RX buffer size, according to example given in documentation. So, we've discarded that solution in favor to fix HAL

Reply:
uartAvailable() is used to check for presence of data.

As for RX buffer size, that is allocated at time of initialization of the uart via IDF API

Reply:
@atanisoft     

So, what do you recommend? add overloaded functions?

Reply:
You could add a new method with the updated signature with the existing one calling into the new method for compatibility. Since this is C code and not C++ I'm not sure if an overloaded method with the same name is possible.

Reply:
Oh, yes. It's C. I'll think how to implement this.